1.x: support Observable.test()

davidmoten opened the pull request:
The new method chaining `TestSubscriber` in 2.x is too good not to have in 1.x as well (while I wait for 2.x to mature and to make my existing unit tests less verbose).

``` java
Observable
  .just(1)
  .test()
  .assertValue(1)
  .assertCompleted();
```

`Observable.test()` returns a `TestSubscriber2` that wraps a `TestSubscriber` and enables method chaining where it can.

I've added tests to get to 100% test coverage though I've been pretty slap-dash about testing some of the details (for example I haven't tested that `awaitXXX` really does wait for stuff).

I've noted it as `@since 1.2.3` but can update to 1.2.2 if you are happy to squeak it into the release.

davidmoten committed:
add TestSubscriber2 to provide method chained version of TestSubscriber and support Observable.test()

akarnokd commented:
I have mixed feelings about this. 2.x required massive amount of unit tests, often with the same structure across the base types where this was a valuable addition. Maybe you could flesh it out in rxjava-extras and provide a function for the `to` operator:

``` java
import static TestConsumer.*;

Observable.just(1)
.to(newTest(0L))
.assertEmpty()
.requestMore(1)
.assertResult(1);
```

You can base it on `Subscriber` and implement `CompletableSubscriber` but I wish we introduced a base interface for SingleSubscriber before going 1.2.

MyDogTom commented:
Just for reference. There are at least two libraries that aim to provide similar functionality (chaining assertions):
- [RxTestWrapper](https://github.com/MyDogTom/RxTestWrapper) - just wraps `TestSubsriber` in a similar manner as this PR
- [rxassertions](https://github.com/ubiratansoares/rxassertions) - provides additional assertions, but also introduce some limitations.

davidmoten commented:
@akarnokd, @MyDogTom 
I have been using this functionality in _rxjava-extras_ using `.to(TestingHelper.test())` for about five weeks. It just seems too useful to miss out on natively (don't have to add extra library to pom.xml, add import, put up with slightly less readable line in code).  Discoverability is sorted for new users too.

For anyone writing tests for their observables I think this is very attractive functionality to have natively and I was stoked to use it with rxjava 2.

Note that my motivation for this is for writing tests in my projects (not rxjava 1.x core, though that would be nice too). 

I forgot to consider `Completable` and `Single` because I don't use them (the majority of our rxjava based code predates their existence). Do they need special treatment (other than addition of a `test` method there)?

akarnokd commented:
How about having an interface with the get and assert methods and the 3 test()s return it. In an internal package then you can implement them and have delegates as necessary.

davidmoten commented:
> How about having an interface with the get and assert methods and the 3 test()s return it. In an internal package then you can implement them and have delegates as necessary.

Yep I can do that.

davidmoten commented:
> How about having an interface with the get and assert methods and the 3 test()s return it. In an internal package then you can implement them and have delegates as necessary.

Having extracted an interface the one implementation can serve all `test()` methods by subscribing to `Completable`, `Single` or `Observable` instances using `subscribe(Subscriber)`. How do you want the implementations to differ?

akarnokd commented on src/main/java/rx/Completable.java:
```diff
@@ -16,19 +16,48 @@
 
-import java.util.*;
-import java.util.concurrent.*;
+import java.util.Arrays;
```
Please make sure you don't unroll `*` imports.

akarnokd commented on src/main/java/rx/observers/TestSubscriber2.java:
```diff
@@ -18,184 +18,70 @@
-        ts.assertValue(value);
-        return this;
-    }
+public interface TestSubscriber2<T> extends Observer<T> {
```
I'd call this `TestConsumer` because that `2` at the end will make people wonder and question endlessly.

akarnokd commented on src/main/java/rx/observers/TestSubscriber2.java:
```diff
@@ -18,184 +18,70 @@
+    void unsubscribe();
     
-}
+    boolean isUnsubscribed();
```
Reinforce with `implements Subscription`?

MyDogTom commented:
I think `test()` should be added also for `BlockingObservable`.

akarnokd commented:
Unnecessary. TestConsumer should have awaitTerminalEvent that blocks.

davidmoten commented:
TestConsumer maybe not ideal because sounds like it's a significantly different beast from a TestSubscriber. I prefer to retain the word Subscriber in the name, what about TestSubscriberFluent?

akarnokd commented:
`AssertableSubscriber`

MyDogTom commented:
+1 for `TestSubscriberFluent` because it reflects the main difference from `TestSubscriber` - fluent API
 -1 for `AssertableSubscriber` it doesn't provide any additional info (`TestSubscriber` is also "assertable").

@akarnokd I still think that `BlockingObservable` should have `test()` method. Let's assume that I encounter some legacy code that use `BlockingObservable` (regardless it's not recommended). I would like to be able write test first and refactor it afterwards (get rid of `BlockingObservable`).

davidmoten commented:
@MyDogTom  puts it well I think, after all `AssertableSubscriber` would provoke as many "what's the difference?" questions as `TestSubscriber2`. `TestSubscriberFluent` describes exactly what the class is.

davidmoten commented:
fixed imports, renamed `TestSubscriber2` to `TestSubscriberFluent` (though I'm not assuming the naming discussion has finished).

squashed commits and rebased on latest 1.x

akarnokd commented:
`AssertableSubscriber` should be an interface with all the relevant `assert` methods that any actual internal test class can implement on top of the respective receiver type. 

> I think, after all AssertableSubscriber would provoke as many "what's the difference?"

If this comes up we will answer that it is the base API interface for asserting a subscriber's state for all 3 reactive types.

I see confusion going the other way around: why `TestSubscriber2` or `TestSubscriberFluent`; its prefix `TestSubscriber` matches, well, `TestSubscriber` itself.

davidmoten commented:
Righto, let's go with `AssertableSubscriber`. Modified code and squashed commits

akarnokd commented on src/main/java/rx/Completable.java:
```diff
@@ -2364,4 +2365,45 @@ public void call() {
+     * @since 1.2.3
+     */
+    @Experimental
+    public final AssertableSubscriber<Void> test(long initialRequestAmount) {
```
`Completable` doesn't need backpressure so this method has no use.

akarnokd commented on src/main/java/rx/Single.java:
```diff
@@ -2675,4 +2676,45 @@ public void call(SingleSubscriber<? super T> singleSubscriber) {
+     * @since 1.2.3
+     */
+    @Experimental
+    public final AssertableSubscriber<T> test(long initialRequestAmount) {
```
Single doesn't support backpressure either.

davidmoten commented:
Thanks @akarnokd, made fixes and squashed commits.
